src/System.Private.CoreLib/src/Internal/Runtime/Augments/EnvironmentAugments.Unix.cs

```diff
@@ -27,10 +27,55 @@ private static void SetEnvironmentVariableCore(string variable, string value)
 
         public static IEnumerable<KeyValuePair<string,string>> EnumerateEnvironmentVariables()
         {
-            if ("".Length != 0)
-                throw new NotImplementedException(); // Need to return something better than an empty environment block.
+            IntPtr block = Interop.Sys.GetEnviron();
+            if (block == IntPtr.Zero)
+                yield break;
 
-            return Array.Empty<KeyValuePair<string,string>>();
+            // Per man page, environment variables come back as an array of pointers to strings
+            // Parse each pointer of strings individually
+            while (ParseEntry(ref block, out string key, out string value))
+            {
+                if (key != null && value != null)
+                    yield return new KeyValuePair<string, string>(key, value);
+
+                // Increment to next environment variable entry
+                block += IntPtr.Size;
+            }
+
+            // Use a local, unsafe function since we cannot use `yield return` inside of an `unsafe` block
+            unsafe bool ParseEntry(ref IntPtr current, out string key, out string value)
```

Comment:
Since you have moved the increment up, this does not need to be `ref  IntPtr current`. It can be just `IntPtr current`